### ISSUE TITLE
Question on Integrator parameters (for documentation)

### ISSUE THREAD

**Issue by jdent02** [bsavery/PRMan-for-Blender#138]
Hi bsavery, I had some in depth questions on some of the integrator parameters.  The Renderman reference docs are good, but they do not sufficiently explain the behavior of several settings.  I was hoping either you or someone you know at Pixar could help me out so I can get it right for this exporter's documentation..........

numIndirectSamples: I'm under the impression that if this is increased past 1, every camera ray will spawn multiple rays at a surface point.  What are the benefits of doing this?  Also, would the bxdf split the rays into different types?  Say for instance the numIndirectSamples is set to 3, then a Disney bxdf with a specular and diffuse component is hit, would there end up being two diffuse rays and one specular ray sent out?  Or vice versa depending on the level of specularity?  The documentation seems to say this is the case, but I wasn't certain.

numDiffuseSamples (and specular, subsurface, etc......): My understanding is that when the sampleMode is set to 'manual' these controls come into effect.  My question is on how the rays spawn.  For every surface that is hit, are all the ray types created, or only the ray types that correspond to the surface?  For instance, you had this scenario:

numDiffuseSamples: 3
numSpecularSamples: 2
numRefractionSamples:3

A camera ray then hits a diffuse surface.  Would only a diffuse ray be spawned, or would all three types spawn?  The documentation makes it seem like the latter case, which if true seems incredibly inefficient and wasteful.

Sorry for the long post.  I feel that any attempt to document these settings should be as thorough as possible, given how complex they are.

**Comment by bsavery** [comment]
Great questions.  And they're actually very related.  The answer to both questions is in the sampleMode setting:

"sampleMode
Controls the strategy for how indirect ray counts are apportioned between diffuse vs. glossy vs. specular Bxdf lobes. Valid values for this parameter are: "bxdf" or "manual". If "sampleMode" is set to "bxdf", then the Bxdf is responsible for balancing the number of indirect rays that will be spawned at each camera hit point between the diffuse vs. glossy vs. specular lobes, and the total number of indirect rays spawned at each camera hit point will be controlled by the numIndirectSamples parameter. If "sampleMode" is set to "manual" the user is responsible for explicitly specifying the number of indirect rays to spawn per Bxdf lobe at each camera hit point via the numDiffuseSamples, numSpecularSamples, numSubsurfaceSamples, and numRefractionSamples parameters. The default is "bxdf"."

When sampleMode is set to bxdf if balances each indirect sample to what it needs (not one of each).  Of course a bxdf like PxrDiffuse will not use the refractive samples though.

In short the answers to your questions are:  
Yes the bxdf balances samples based on certain factors.  So it might be that 3 indirect samples = 2 spec 1 diffuse

If a bxdf doesn't have certain lobes and those samples are set, they are not used.

**Comment by jdent02** [comment]
Perfect.  I assume the line of thinking is that individual diffuse/specular/etc.. rays are cheaper than camera rays to compute?  Cycles has a mode that operates on a similar principle and I believe it's the way Arnold works.

**Comment by bsavery** [comment]
Diffuse/specular rays can be cheaper.  Probably cheaper is the wrong word.  But in a nutshell yes.

**Comment by jdent02** [comment]
Thanks for clearing that up.  It sounds like the options might be more trouble than they're worth.....
